packages/web/src/common/store/tipping/selectors.ts

```diff
@@ -31,3 +40,143 @@ export const getTipToDisplay = (state: CommonState) =>
   state.tipping.tipToDisplay
 export const getShowTip = (state: CommonState) => state.tipping.showTip
 export const getMainUser = (state: CommonState) => state.tipping.mainUser
+
+const getOptimisticSupportingBase = (state: CommonState) => {
+  const { supporting, supportingOverrides } = state.tipping
+
+  /**
+   * Copy the supporting map into the eventually-merged map.
+   */
+  let mergedMap: SupportingMap = {}
+  Object.assign(mergedMap, supporting)
+
+  /**
+   * Merge the default and override maps.
+   */
+  const userIds = (Object.keys(supportingOverrides) as unknown) as ID[]
+  for (const userId of userIds) {
+    // If the supporting map for a given user id exists in the overrides
+    // but not in the default, copy the override map into the merged map.
+    const shouldOverrideMap = !supporting[userId]
+    if (shouldOverrideMap) {
+      mergedMap = {
+        ...mergedMap,
+        [userId]: supportingOverrides[userId]
+      }
+    } else {
+      // If the supporting value for a given user id and receiver id exists
+      // in the overrides but not in the default,
+      // OR
+      // if the existing value in the default map has a smaller amount
+      // than that in the override, the update default value with the
+      // override value
+      const receiverIds = (Object.keys(
+        supportingOverrides[userId]
+      ) as unknown) as ID[]
+      for (const receiverId of receiverIds) {
+        const shouldOverrideReceiver =
+          !supporting[userId][receiverId] ||
+          stringWeiToBN(supporting[userId][receiverId].amount).lt(
+            stringWeiToBN(supportingOverrides[userId][receiverId].amount)
+          )
+        if (shouldOverrideReceiver) {
+          mergedMap = {
+            ...mergedMap,
+            [userId]: {
+              ...mergedMap[userId],
+              [receiverId]: supportingOverrides[userId][receiverId]
+            }
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Re-rank everything based on newly merged map.
+   */
+  let result: SupportingMap = {}
+  const mergedUserIds = (Object.keys(mergedMap) as unknown) as ID[]
+  for (const userId of mergedUserIds) {
+    result = {
+      ...result,
+      [userId]: optimisticallyUpdateSupporting(mergedMap[userId])
+    }
+  }
+
+  return result
+}
+export const getOptimisticSupporting = createSelector(
+  getOptimisticSupportingBase,
+  result => result
+)
+
+const getOptimisticSupportersBase = (state: CommonState) => {
```

Comment:
These two functions look very similar, can we break out their individual pieces into reusable functions? Might have to use generics to make typescript happy?